xenvm/vmact.ml

```diff
@@ -904,11 +904,27 @@ let shutdown_vm xc xs xal state force reason =
 			 * to another asynchronous shutdown happening *)
 			if not (Xal.domain_is_dead xal domid) then raise e);
 
-		info "vm acknowledged shutdown request. waiting release";
-		(* wait domain release with xal *)
+			info "vm acknowledged shutdown request. waiting release";
+			(* wait domain release with xal *)
+		let retry = 
+			(for i = 0 to 5 do
+				let retry_result = 
+					(try
+						Xal.wait_release xal ~timeout:(30.) domid; true
+					with Xal.Timeout ->
+						info "No shutdown after 30 seconds, retrying attempt %d of 5" i;
+						if force then (
+							Domain.hard_shutdown ~xc domid reason; true
+						) else (
+							Domain.shutdown_ack ~timeout:30. ~xc ~xs domid reason
```

Comment:
Xal.wait_release checks to see if the domain is dead for duration of timeout (30 sec). If we reached the timeout, try issuing the shutdown command again.  Ocaml doesn't have 'break' so if the domain is dead, wait_release returns immediately and no additional shutdowns are issued, but the loop is run to completion.